selectors/selectors.md

```diff
@@ -302,67 +265,152 @@ If it's a file in some directory, you can also start at a deeper level:
+type ExploreConditional struct {
+	condition Condition (alias "&")
+	next Selector (alias ">")
 }
 
-# SelectRecursive applies a selector to this node, then in addition traverses
-# to any nodes matched and applies itself again to that node.  A depth counter
-# is decremented with each recursion to ensure the process terminates.
-type SelectRecursive struct {
-	depthLimit Int (alias "d")
-	next Selector (alias ":")
-	cidLimit Link
+## Matcher marks a node to be included in the "result" set.
+## (All nodes traversed by a selector are in the "covered" set (which is a.k.a.
+## "the merkle proof"); the "result" set is a subset of the "covered" set.)
+##
+## In libraries using selectors, the "result" set is typically provided to
+## some user-specified callback.
+##
+## A selector tree with only "explore*"-type selectors and no Matcher selectors
+## is valid; it will just generate a "covered" set of nodes and no "result" set.
+type Matcher struct {
+	onlyIf optional Condition # match is true based on position alone if this is not set.
+	label optional String # labels can be used to match multiple different structures in one selection.
 }
 
-# SelectUnion is a union of other selectors: if any of the selectors says
-# a node matches, it matches; and if any of the selectors return a selector
-# for a pathsegment, it'll be used; and if more than one selector returns
-# a selector for a pathsegment, a new union selector will be implicitly
-# generated during evaluation to carry all of them through.
-type SelectUnion list [Selector]
+## Condition is expresses a predicate with a boolean result.
+##
+## Condition clauses are used several places:
+##   - in Matcher, to determine if a node is selected.
+##   - in ExploreRecursive, to halt exploration.
+##   - in ExploreConditional,
+##
+##
+## TODO -- Condition is very skeletal and incomplete.
+## The place where Condition appears in other structs is correct;
+## the rest of the details inside it are not final nor even completely drafted.
+type Condition union {
+	# We can come back to this and expand it later...
+	# TODO: figure out how to make this recurse correctly, so I can say "hasField{hasField{or{hasValue{1}, hasValue{2}}}}".
+	| Condition_HasField "hasField"
+	| Condition_HasValue "=" # will need to contain a kinded union, lol.  these conditions are gonna get deep.)
+	| Condition_HasKind "%" # will ideally want to refer to the DataModel ReprKind enum...!  will we replicate that here?  don't want to block on cross-schema references, but it's interesting that we've finally found a good example wanting it.
+	| Condition_IsLink "/" # will need this so we can use it in recursions to say "stop at CID QmFoo".
+	| Condition_GreaterThan "greaterThan"
+	| Condition_LessThan "lessThan"
+	| Condition_And "and"
+	| Condition_Or "or"
+	# REVIEW: since we introduced "and" and "or" here, we're getting into dangertown again.  we'll need a "max conditionals limit" (a la 'gas' of some kind) near here.
```

Comment:
Condition_Not? Just saying you need And/Not/Or to derive all boolean logic.

Reply:
Probably :)  I kinda just spitballed this section, in order to say "something roughly like this belongs here".  It almost certainly needs another round of draft -- would definitely give the `Condition` and its members another hard design look first before trying to implement these.

(Also figure they're probably roughly one of the last things to do in the overall order of implementing the selector system.)

Reply:
ok. Imma hope i get go aways with filecoin before I need conditionals.